lms/djangoapps/grades/models.py

```diff
@@ -0,0 +1,253 @@
+        sorted_blocks = sorted(
+            blocks,
+            key=lambda block: '{}{}{}'.format(block.block_key, block.max_score, block.weight)
+        )
+        _blocks_json = json.dumps(
+            [
+                block.to_dict()
+                for block in sorted_blocks
+            ]
+        )
+        hashed = sha256(_blocks_json).hexdigest()
+        model, _ = cls.objects.get_or_create(hashed=hashed, defaults={'_blocks_json': _blocks_json})
+        return model
+
+    @property
+    def blocks(self):
+        """
+        Returns the blocks_json data stored on this model as an list of BlockRecords. If for some reason block_json
+        is not parsable, a json error will bubble up.
+        """
+        block_dicts = json.loads(self._blocks_json)
+        return [BlockRecord(data['weight'], data['max_score'], data['block_key']) for data in block_dicts]
+
+    @blocks.setter
+    def blocks(self, value):
+        """
+        Not implemented, as VisibleBlocks instances are intended to be write-once and not change after creation.
+        """
+        raise NotImplementedError(
+            "Property 'blocks' cannot be modified on an existing VisibleBlocks model. Create a new instance."
+        )
+
+
+class BlockRecord(object):
+    """
+    An object encapsulating all relevant information for representing a block at the time it was used for grade
+    calculation.
+    """
+
+    def __init__(self, weight, max_score, locator):
+        """
+        Creates a BlockRecord.
+
+        Params:
+            weight (float)
+            max_score (float)
+            locator (BlockUsageLocator or string)
+        """
+        self.weight = weight
+        self.max_score = max_score
+        # pylint: disable=protected-access
+        self.block_key = locator._to_string() if isinstance(locator, BlockUsageLocator) else locator
+
+    def to_dict(self):
+        """
+        Serialize this object to a dict object.
+        """
+        return {
+            'weight': self.weight,
+            'max_score': self.max_score,
+            'block_key': self.block_key,
+        }
+
+    @property
+    def locator(self):
+        """
+        Convenience property allowing a BlockUsageLocator to be returned from this object's serialized data.
+        """
+        return BlockUsageLocator._from_string(self.block_key)  # pylint: disable=protected-access
+
+
+class PersistentSubsectionGrade(TimeStampedModel):
+    """
+    A django model tracking persistent grades at the subsection level.
+
+    TODO: here are the other query patterns listed in the ticket that are not currently used. Should any of these
+    indices be built?
+        user_id, course_id, content_type, is_valid
+        course_id, edit-timestamp
+        edit-timestamp
+    """
+
+    class Meta(object):
+        index_together = [
+            ('user_id', 'usage_key')
+        ]
+
+        unique_together = (('user_id', 'usage_key'))
```

Comment:
Course keys are not attached to usage keys in the case of old mongo courses (usage keys that start with `i4x`) -- it's why we have all the crazy `map_into_course()` calls littered across the code. Also, when rendering the progress page, a very natural query would be "give me all the subsection grades for this user in this course". Relying on prefix matching for that may get hairy -- CCX course IDs have a different starting prefix and an extra parameter (e.g. **ccx-v1**:edX+10S+2015_Split+ccx**@1**), and other course types may introduce different conventions. The easiest way to avoid odd bugs is to just treat it as opaquely as possible and store it as a separate field.